### PR TITLE
[POAE7-2320] Support string as key/target in group by

### PR DESCRIPTION
**Pull request by Deegue**
### What changes were proposed in this pull request?
Enable string support in Group by and add some tests.

### Why are the changes needed?
Improvement.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UTs.

### Which label does this PR belong to?
cider, test

### CODE REVIEW

File: cider/exec/template/GroupByAndAggregate.cpp [hunk 1]
```diff
@@ -323,21 +323,7 @@ GroupByAndAggregate::GroupByAndAggregate(
...
-    if (groupby_ti.is_bytes()) {
-      throw std::runtime_error(
-          "Cannot group by string columns which are not dictionary encoded.");
-    }
-    if (groupby_ti.is_buffer()) {
-      throw std::runtime_error("Group by buffer not supported");
-    }
-  }
-}
+    , group_cardinality_estimation_(group_cardinality_estimation) {}
```
Comment (winningsix) [inline cider/exec/template/GroupByAndAggregate.cpp:326]: Move this to header this we do nothing in constructor.

File: cider/tests/functionality/CiderGroupByTest.cpp [hunk 2]
```diff
@@ -48,6 +48,29 @@
     }                                                                                  \
   };
 
+class CiderGroupByVarcharTest : public CiderTestBase {
+ public:
+  // TODO(yizhong): make col_d to 50% chance nullable and change min length to 0
+  // after null string is supportedz
```
Comment (winningsix) [inline cider/tests/functionality/CiderGroupByTest.cpp:54]: typo: supportedz

**Review by winningsix — COMMENTED** [review]
Hmm, there is nothing else we need to and just need to enable it by removing error throwing, right?

**Comment by winningsix** [comment]
@bigPYJ1151 please help review this.

File: cider/exec/template/GroupByAndAggregate.cpp (hunk 1 above)
Comment (jikunshang) [reply cider/exec/template/GroupByAndAggregate.cpp:326]: No need to move. an empty constructor is a constructor impl, it's good to separate define and impl in .h and .cpp file.

File: cider/tests/functionality/CiderGroupByTest.cpp [hunk 3]
```diff
@@ -440,6 +467,128 @@ TEST_F(CiderGroupByPrimitiveTypeMixTest, noConditionGroupByMultiColTest) {
       true);
 }
 
+TEST_F(CiderGroupByVarcharTest, varcharGroupByTest) {
+  /*one not null varchar group by key*/
+  assertQuery(
```
Comment (jikunshang) [inline cider/tests/functionality/CiderGroupByTest.cpp:472]: can you add some simple cases like this ? `SELECT SUM(col_a), col_c FROM table_test GROUP BY  col_c`

**Comment by Deegue** [comment]
> Hmm, there is nothing else we need to and just need to enable it by removing error throwing, right?

Yes, we only need to enable it here since we have finished [[POAE7-2277]](https://github.com/intel-innersource/frameworks.ai.modular-sql.velox-plugin/pull/301) and [[POAE7-2278]](https://github.com/intel-innersource/frameworks.ai.modular-sql.velox-plugin/pull/318).

File: cider/tests/functionality/CiderGroupByTest.cpp (hunk 3 above)
Comment (Deegue) [reply cider/tests/functionality/CiderGroupByTest.cpp:478]: added

File: cider/tests/functionality/CiderGroupByTest.cpp (hunk 2 above)
Comment (Deegue) [reply cider/tests/functionality/CiderGroupByTest.cpp]: done

File: cider/tests/functionality/CiderGroupByTest.cpp [hunk 4]
```diff
@@ -48,6 +48,29 @@
...
+        "CREATE TABLE table_test(col_a BIGINT NOT NULL, col_b BIGINT NOT NULL, col_c "
+        "VARCHAR NOT NULL, col_d VARCHAR);";
+    input_ = {std::make_shared<CiderBatch>(
+        QueryDataGenerator::generateBatchByTypes(500,
+                                                 {"col_a", "col_b", "col_c", "col_d"},
+                                                 {CREATE_SUBSTRAIT_TYPE(I64),
+                                                  CREATE_SUBSTRAIT_TYPE(I64),
+                                                  CREATE_SUBSTRAIT_TYPE(Varchar),
+                                                  CREATE_SUBSTRAIT_TYPE(Varchar)},
+                                                 {0, 0, 0, 0},
```
Comment (yma11) [inline cider/tests/functionality/CiderGroupByTest.cpp:67]: Don't you need set correct null_chance based on your defined schema?

Comment (Deegue) [reply cider/tests/functionality/CiderGroupByTest.cpp:67]: Commented a TODO at Line53, let me change the `null_chance` after null string is supported.

File: cider/tests/functionality/CiderGroupByTest.cpp [hunk 5]
```diff
@@ -20,7 +20,7 @@
  */
 
 #include <gtest/gtest.h>
-#include "tests/utils/CiderTestBase.h"
+#include "CiderTestBase.h"
```
Comment (xieqi) [inline cider/tests/functionality/CiderGroupByTest.cpp:23]: as a practice, better to include the full file path, see [Names and Order of Includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

Comment (jikunshang) [reply cider/tests/functionality/CiderGroupByTest.cpp:23]: let's add `include_order` check in `cpplint` later.

**Review by jikunshang — APPROVED** [review]

**Comment by Deegue** [comment]
Gentle ping @winningsix @yma11 for another review since this PR blocks many TPC-H queries like Q1, Q5, Q18 according to offline discussion with @harborn .

**Review by winningsix — APPROVED** [review]

**Pull request merged**